File: Problems/66. Plus One/66.py
test_cases = [
    [1, 2, 3],
    [4, 3, 2, 1],
    [9],
    [5, 9, 9],
    [9, 9],
    [9, 8, 9]
]

expected_results = [
    [1, 2, 4],
    [4, 3, 2, 2],
    [1, 0],
    [6, 0, 0],
    [1, 0, 0],
    [9, 9, 0]
]


class Solution:
    def plusOne(self, digits):
        for i in range(len(digits)-1, -1, -1):
            if digits[i] == 9:
                digits[i] = 0
            else:
                digits[i] += 1
                return digits
        return [1] + digits


        # Testing
for i in range(0, len(test_cases)):
    print("==================================")
    result = Solution.plusOne(
        Solution, test_cases[i])
    if result == expected_results[i]:
        print(f"==Test case {i+1} was valid")
        print(f"✔️  Got '{result}', expected {expected_results[i]}")
    else:
        print(f"==Test case {i+1} was INVALID")
        print(f"❌ Got '{result}', expected {expected_results[i]}")

File: Problems/66. Plus One/66.rs
fn plus_one(digits: &mut Vec<i32>) -> Vec<i32> {
    for i in (0..digits.len()).rev() {
        if digits[i] == 9 {
            digits[i] = 0;
        } else {
            digits[i] += 1;
            return digits.to_vec();
        }
    }
    digits.insert(0, 1);
    digits.to_vec()
}

fn main() {
    let test_cases = vec![
        vec![1, 2, 3],
        vec![4, 3, 2, 1],
        vec![9],
        vec![5, 9, 9],
        vec![9, 9],
        vec![9, 8, 9]
    ];
    let expected_results = vec![
        vec![1, 2, 4],
        vec![4, 3, 2, 2],
        vec![1, 0],
        vec![6, 0, 0],
        vec![1, 0, 0],
        vec![9, 9, 0]
    ];

    for i in 0..test_cases.len() {
        println!("==================================");
        let result = plus_one(&mut test_cases[i]);
        if result == expected_results[i] {
            println!("==Test case {} was valid", i+1);
            println!("Got '{:?}', expected {:?}", result, expected_results[i]);
        } else {
            println!("==Test case {} was INVALID", i+1);
            println!("Got '{:?}', expected {:?}", result, expected_results[i]);
        }
    }
}

